content/en/blog/_posts/2023-03-08-memory-qos-cgroups-v2/index.md

```diff
@@ -0,0 +1,260 @@
+       <br>
+       <br>
+       <i>We use a formula to calculate <code>memory.high</code>, depending on container's memory request, memory limit or node allocatable memory (if container's memory limit is empty) and a throttling factor. Please refer to the KEP for more details on the formula.</i>
+       </td>
+   </tr>
+</table>
+
+**Note**: If a container has no memory limit specified, we substitute limits.memory for node allocatable memory.
+
+### Support for Pod QoS classes
+Based on user feedback for the Alpha feature in Kubernetes v1.22, some users would like 
+to opt out of MemoryQoS on a per-pod basis to ensure there is no early memory throttling.
+Therefore, in Kubernetes v1.27 Memory QOS also supports memory.high to be set as per 
+Quality of Service(QoS) for Pod classes. Following are the different cases for memory.high
+as per QOS classes:
+
+1. **Guaranteed pods** by their QoS definition require memory requests=memory limits and are 
+not overcommitted. Hence MemoryQoS feature is disabled on those pods by not setting 
+memory.high. This ensures that Guaranteed pods can fully use their memory requests up
+to their set limit, and not hit any throttling.
+
+2. **Burstable pods** by their QoS definition require at least one container in the Pod with
+CPU or memory request or limit set.
+    
+    * When requests.memory and limits.memory are set, the formula is used as-is:
+    ![](./container-memory-high-limit.svg)  
+    <sub>i: the i<sup>th</sup> container in one pod</sub>
+
+    * When requests.memory is set, limits.memory is not set, we substitute limits.memory
+    for node allocatable memory in the formula:
+    ![](./container-memory-high-no-limits.svg)  
+    <sub>i: the i<sup>th</sup> container in one pod</sub>
+
+3. BestEffort The pod gets a BestEffort class if limits.memory and requests.memory are not 
+set. We set requests.memory = 0 and substitute limits.memory for node allocatable memory 
+in the formula:
+
+    ![](./container-memory-high-best-effort.svg)  
+    <sub>i: the i<sup>th</sup> container in one pod</sub>
+
+### Cgroup Hierarchy
+When container memory requests are made, kubelet passes `memory.min` to the back-end 
+CRI runtime (possibly containerd, cri-o) via the `Unified` field in CRI during 
+container creation. The `memory.min` in container level cgroup will be set to:  
+
+![](./container-memory-min.svg)  
+<sub>i: the i<sup>th</sup> container in one pod</sub>
+
+Since the `memory.min` interface requires that the ancestor cgroup directories are all 
+set, the pod and node cgroup directories need to be set correctly. 
+
+`memory.min` in pod level cgroup:  
+![](./pod-memory-min.svg)  
+<sub>i: the i<sup>th</sup> container in one pod</sub>
+
+`memory.min` in node level cgroup:  
```

Comment:
What if the container runtime isn't running Pods and sandboxes in the root cgroup? (I don't know if we support this). I'd imagine the sysadmin then has to set `memory.min` appropriately outside of Kubernetes.

Reply:
The MemoryQoS feature will enforce `memory.min` on the node level kubepods cgroup (see [here](https://github.com/kubernetes/kubernetes/blob/330b5a2b8dbd681811cb8235947557c99dd8e593/pkg/kubelet/cm/node_container_manager_linux.go#L150).

This cgroup is managed by the kubelet and all pods are children of it, so it's not possible to create pods that run outside of it (unless you are running containers directly via the container runtime unmanaged by kubelet).

Reply:
Is this right?
```suggestion
`memory.min` in node level cgroup (for example, `kubepods.slice` if you use the systemd driver):
```